### ISSUE TITLE
Insert to mysql, can find data inserted in mysql but the method returns with Err(RecordNotFound("Failed to find inserted item")).

### ISSUE THREAD

**Issue by westYY123** [SeaQL/sea-orm#2139]
## Description

hi, I am developing the backend of an app with axum and sea-orm, so there is a route to register a user, and insert the user to databse, i am using mysql as my database, but any time i call this route, it returns with Err(RecordNotFound("Failed to find inserted item")), however, i can find data in my database.

## Steps to Reproduce

the below is the router 

```
pub async fn register(
    State(app_data): State<Arc<AppData>>,
    Json(req): Json<RegisterRequest>,
) -> AppResult<Json<RegisterResponse>> {
    let res = user::insert_user(&app_data.mysql_client, req.username, req.password).await;
    println!("{res:?}");
    // .map_err(|_| AppError::InternalError)?;
    Ok(Json(RegisterResponse { success: true }))
}
```
the below is the database operator, the res is Err(RecordNotFound("Failed to find inserted item")).
```
pub async fn insert_user(
    conn: &DatabaseConnection,
    username: String,
    password: String,
) -> AppResult<Model> {
    let existing_user = get_user(conn, &username)
        .await
        .map_err(|_| AppError::InternalError)?;
    match existing_user {
        Some(_) => Err(AppError::ExistingSameUsername),
        None => {
            let user = ActiveModel {
                id: Set("test-id".to_string()),
                username: Set(username),
                password: Set(password),
            };

            let res = user.insert(conn).await;
            println!("{res:?}");
            res.map_err(|_| AppError::InternalError)
        }
    }
}

```

### Expected Behavior

returns the Model.

### Actual Behavior

returns err Err(RecordNotFound("Failed to find inserted item"))

### Reproduces How Often

every time

## Versions

sqlx = "0.7"
sea-orm = { version = "0.12", features = [
    "sqlx-mysql",
    "runtime-tokio-native-tls",
    "macros",
    "with-chrono",
] }

**Comment by zawata** [comment]
I encountered this bug yesterday and spent a few hours troubleshooting it

Tl;Dr: Either use `Entity::insert(user).exec(conn)` or make the table's primary key AUTO_INCREMENT.

The `ActiveModel.insert` function attempts to return the freshly-inserted row if it was told that the insert was successful. If the database supports returning inserted rows then it will convert the row result to a model and return it. If it doesn't, it uses the "last_insert_id" field of the result to look up the row it just inserted. MySQL doesn't set the last_insert_id on the insert result unless the primary key of the table is auto_increment:
https://dev.mysql.com/doc/refman/8.0/en/information-functions.html#function_last-insert-id

but sea_orm doesn't account for this, assuming that the last_insert_id is valid:
https://github.com/SeaQL/sea-orm/blob/010433bf1633dd387f8493071c0b3838a7fbb575/src/executor/insert.rs#L295

Given that there's separate `insert_with_returning` and `insert_without_returning` functions, I have to assume this is expected behavior but I still wish this function could handle this error more gracefully or more obviously, or at least document it somehow/somewhere?

The simple fix is to use `Entity::insert(user).exec(conn)` which instead returns the insert result, which, if `Ok`, contains `last_insert_id` but it's set to 0. Doesn't really matter since you know the insert succeeded and you should already have the ID you set in the active model so you can just follow it up with a `Entity::find_by_id(id).one(conn)` to get the fresh model.

this is what I'm doing: 

```rust
  let new_user_id = rand::random::<u64>();

  let user = ActiveModel {
      id: Set(new_user_id),
      email: Set(email.clone()),
      password_hash: Set(password_hash.clone())
  };

  Entity::insert(user).exec(conn).await?;
  let new_user = Entity::find_by_id(new_user_id).one(conn).await?.ok_or(anyhow!("inserted user not found"))?;

  return Ok(new_user);
```

The more complex fix is to make sure your primary key column is set to auto-increment so mysql will actually return the inserted ID.

or just switch to postgres which doesn't seem to have this issue.

**Comment by lcmgh** [comment]
same issue on postgres

**Comment by DMaxter** [comment]
I found that the SeaORM doesn't really support databases that don't support the `RETURNING` clause.
I hit this error in MySQL after trying to fix another error I found which is due to bad implementation for MySQL specifically.

For MySQL (which doesn't support the `RETURNING` clause), we hit the code here https://github.com/SeaQL/sea-orm/blob/a50c848f7d634233c78148e3bb1c8e82b5f448ba/src/executor/insert.rs#L372-L381

But looking at it, we can see 2 errors:
* On line 381, it is trying to transform whatever is the `last_id` (which is not returned) into a u64 (which it might not be, because primary key can be of any type). But for this case it is 0, so not really a big problem
* On line 378, it is checking if it is MySQL database and if it is and if the `last_id` is 0 (which as we saw, it always is). If both conditions are met, it just returns `RecordNotInserted`, even though we know a record **was** inserted because the number of affected rows is greater than 0, as we already past the condition https://github.com/SeaQL/sea-orm/blob/a50c848f7d634233c78148e3bb1c8e82b5f448ba/src/executor/insert.rs#L369-L371

After solving this, I hit the same issue as you did and got to the same conclusions as @zawata .

However, I think it is possible to still return the object we just inserted, as we have the primary key which is the id, so we are able to search with it, we can just ignore `last_id`. I'm still trying to figure out how, because there is no direct conversion from the type of `primary_key` to the type the `find_by_id` function expects

**Comment by DMaxter** [comment]
I'm attaching a very simple reproducer, that only needs the DB name and credentials changed. The table is as simple as 3 strings, where `email` is the primary key

[reproducer.zip](https://github.com/user-attachments/files/19145585/reproducer.zip)

**Comment by DMaxter** [comment]
Well, after a different hunch, I was able to make this work. Looks like I was getting this error because as stated in the documentation
> By default, auto_increment is implied for `primary_key` column. Override it by specifying `false`."

So, in my reproducer, after changing `sea_orm(primary_key)` to `sea_orm(primary_key, auto_increment = false)` everything seemed to start working.

Because by default is is enabled, and if our primary key is neither composite (to disable it) or a number, it will fail to get a value and so we match the code section mentioned previously.

**Comment by DMaxter** [comment]
However, I think the error should be different as it is misleading

**Comment by DMaxter** [comment]
Same with me. Auto increment should be false if type is different from integer and also the error should be more explicit

**Comment by holmofy** [comment]
same issue on postgres

**Comment by paramako** [comment]
same

**Comment by Huliiiiii** [comment]
The issue seems to be different in PostgreSQL than in MySQL. 
Can you provide more information to help me pinpoint the issue?

**Comment by tyt2y3** [comment]
> > By default, auto_increment is implied for `primary_key` column. Override it by specifying `false`."
> 
> So, in my reproducer, after changing `sea_orm(primary_key)` to `sea_orm(primary_key, auto_increment = false)` everything seemed to start working.
> 
> Because by default is is enabled, and if our primary key is neither composite (to disable it) or a number, it will fail to get a value and so we match the code section mentioned previously.

thank you, this is the right solution. 

> Auto increment should be false if type is different from integer

sadly this design can't handle the case when using a wrapped primary key type, i.e. `struct MyId(pub i32)`
the derive macro cannot determine whether a custom type can support auto-increment or not

**Comment by Huliiiiii** [comment]
In #2881, I made String and Uuid non-auto-incrementing by default, which should fix the issue in most cases.
